### ISSUE TITLE
Orbot support

### ISSUE THREAD

**Issue by JasperWeiss** [theScrabi/NewPipe#39]
Heya, Mr tinfoil hat wearing privacy guy here.
Could you (perhaps later down the road) add support for orbot? Even though the app is open-source, Google is still receiving querries from your IP and so it'd be nice if newpipe supports orbot.
After all, the main reason of not using Gapps is to avoid Google's data mining.

I'd recommend a dropdown menu "orbot support" "disabled,enabled,forced"
Where forced will not allow a connection to be made unless orbot is running.

**Comment by darkon5** [comment]
Doesn't Orbot override every connection that happens on the phone and push it through local proxy server into the Tor network? It should do it already then.

I don't know, I don't use Orbot since Android 2.3.4.

**Comment by JasperWeiss** [comment]
Not by default. You can use it as an VPN or enable transparant proxy but the latter requires root. That said, doing so is rather stupid since that makes every app connect over tor; revealing your identity. (Google play store, whatsapp etc know your real identity so it can be linked to the IP your comming from) there is a reason you shouldn't login over tor.

**Comment by theScrabi** [comment]
Hm the problem is –as far as i observed it– youtube only streams to devices witch have the IP that is given in the request URL. Is that true?

**Comment by JasperWeiss** [comment]
Nah, bassically all requests go through the tor network. Youtube gets the request from the tor exit node and sends back the response, the exit node then passes it on to the next node all the way down to you. I'll see if I can find you some example code

**Comment by theScrabi** [comment]
Ok. Yep help is good.

**Comment by JasperWeiss** [comment]
It seems there's a library from the guardianpoject itself that will allow you to intergrate tor support

(But by all means, finish the UI and other important stuff first. Tor support can wait for now. It should be pretty easy to implement, though)

https://github.com/guardianproject/NetCipher

**Comment by theScrabi** [comment]
... unless someone helps me :)

**Comment by JasperWeiss** [comment]
I would like to help, it's just that I've no understanding of programming and I've no idea how to get started. I usually open an empty project thinking "now what do I do?"
How did you learn all of this?

**Comment by theScrabi** [comment]
Well i knew programming with Qt C++. Then earlier this year i bought a book about android programming. I'm not the best programmer, but I'd say android is not the platform of choice if you want to learn/start programming. Hm...

**Comment by JasperWeiss** [comment]
Do you mind if I look through the newpipe code and play around with for learning purposes?
I did some c++ coding myself so Java shouldn't be that hard the learn

**Comment by theScrabi** [comment]
Hm... Well try it. :) Good luck.

**Comment by gjedeer** [comment]
I tried approaching this problem using NetCipher and while adding a setting and patching Downloader.java was easy, it's not that easy when it comes to video stream.

Android's MediaPlayer has [no proxy support](https://stackoverflow.com/questions/10627278/how-to-set-proxy-on-mediaplayer) and it requires a [hacky workaround](https://code.google.com/p/npr-android-app/source/browse/Npr/src/org/npr/android/news/StreamProxy.java?spec=svnce46f6fe92353676bf9d98b530fceb8328aceb85&r=e4984187f45c39a54ea6c88f71197762dbe10e72) to work.

**Comment by gjedeer** [comment]
It seems that Google themselves found the MediaPlayer API too limited and created a more sophisticated alternative, [ExoPlayer](https://github.com/google/ExoPlayer). It does support custom data sources and it seems that the implementation is [quite simple](https://google.github.io/ExoPlayer/doc/reference/com/google/android/exoplayer/upstream/DataSource.html). But then, moving New Pipe from MediaPlayer to ExoPlayer is probably a huge project.

**Comment by theScrabi** [comment]
No I don't think so.
The part where NewPipe uses the MediaFramework is pretty small xD. It's probably less than 100 lines of code. Therefore the hardest part would be to import exoplayer to the project.

Here you  go. This should be the only part where NP uses the MediaFramework: https://github.com/theScrabi/NewPipe/blob/master/app/src/main/java/org/schabi/newpipe/PlayVideoActivity.java

**Comment by medavox** [comment]
I'll also need to switch the BackgroundPlayer code, but I've been away the last few days. Hopefully I'll make some progress on it this week.

**Comment by PacoBell** [comment]
Not that I object to the idea of streaming video over Tor, conceptually, but if people start doing this en masse without each additional user also setting up a new, permanent exit node, it is likely to DoS an already strained network. It's a tragedy of the commons, if you will.

**Comment by JasperWeiss** [comment]
Hmm valid point. Personally I'd be willing to run a relay. But an exit node seems risky since that will result in all sorts of sketchy stuff comming from my network

**Comment by gjedeer** [comment]
I actually mentioned youtube streaming over Tor on one of the Tor
Project lists - in the context that it's bad - and the reply was "why
do you think it's bad?".

So don't concern yourself with that.

About my patch moving to exoplayer: not happening any time soon, sorry,
have personal stuff to handle.

**Comment by theScrabi** [comment]
Ok no problem. Take your time.

About Tor: Right downloading videos via Tor slows their network down. So we should think of not implementing this for streaming. But what could be done over Tor is all the crawling stuff.

**Comment by gjedeer** [comment]
https://github.com/gjedeer/NewPipe/tree/tor <- this is metadata over tor, it should work.

**Comment by theScrabi** [comment]
@gjedeer Hm I've tried your branch, and yes it works nice with orbot, and also bypasses streaming. But also I looked into your code, and found several issues we need to fix before we can finally merge. Please had over to our [slack page](http://invite.chschtsch.ml/) so we can work on this :)

**Comment by gjedeer** [comment]
Yeah, I know it's not ready to merge. I don't know when I'll be able to
work on this again so I just dumped what I did a few weeks ago.

**Comment by theScrabi** [comment]
Well it locked pretty solid to me. The only think that did not work was that you couldn't switch it of. But don't hurry take your time :)

**Comment by JasperWeiss** [comment]
It sure works nice, though I wonder if exluding the video stream from the tor proxy would defeat the purpose. Since the actual video that is watched is still in the clear.
One reason why proper orbot support is the fact that watching youtube on the tor browser bundle is insecure as it requires javascripts. Newpipe could be a means to securely browse youtube in opressed countries. 
There's plenty of people that watch youtube over tor and I think that 1 youtube app with orbot support won't make much of a difference. People who'd use it would watch it on their desktop anyway.
Any thoughts?

**Comment by theScrabi** [comment]
Hm maybe we could split this up and let the user deside via setting. So either you crawl over tor, and bypass streaming, or you do both over tor. We could than add a notice telling the user that using tor for streaming might slow the network down.

**Comment by gjedeer** [comment]
Ask the network operators before jumping to such conclusions. I'm pretty sure the last time I asked they were OK with streaming videos over Tor. The network has enough exit node bandwidth now, and more nodes are needed mostly for diversity, AFAIK.

By the way, would be nice to add regular proxy support too, for people who don't want to go full anon and just bypass regional restrictions. Should be a matter of adding a few more settings in the prefs.

**Comment by JasperWeiss** [comment]
I like the way lightningbrowser implemented it. Bassically you tab proxy and a window pops up with a few options like tor, i2p, none and manual where if you tap manual it opens a smaller window on top with the adress and port.

https://play.google.com/store/apps/details?id=acr.browser.lightning

https://f-droid.org/app/acr.browser.lightning

**Comment by theScrabi** [comment]
Right Lightning is pretty awsome. About asking the Operaters. Im at 32c3 right now. I could ask them personally xD.

**Comment by JasperWeiss** [comment]
Great idea, ask them :)

**Comment by theScrabi** [comment]
When i find them xD

**Comment by gjedeer** [comment]
Hey, dude, was that you in the MicroG meeting mentioning New Pipe?

**Comment by theScrabi** [comment]
xD yap. That was me :)
About tor, later on I asked R.Dingledine if it was ok. Now it seems to be a bit more complicated. He said, streaming was generally OK if we didn't have to much users who do that. So he said it would be good for us to know how much are out there who use NewPipe and would be willing to use it over tor.
@gjedeer If you are still at 32c3 tomorrow, we could talk about that, so if you like come over to our slack group.

**Comment by eighthave** [comment]
There are not yet a ton of users of NewPipe, so I think streaming over Tor should be not be a problem.  If Google suddenly made YouTube default to using Tor, that would be a problem.  I'm working on this now, I'll have something to submit by the end of tonight.

**Comment by eighthave** [comment]
I just found another thing that'll have to be handled: downloading.  It currently uses `Context.DOWNLOAD_SERVICE` so that won't go through Tor.  It is easy enough to make a custom downloader, so that's the thing to do for Tor support.

**Comment by theScrabi** [comment]
gjedeer already started working on this. You may only have to finish his work.

**Comment by eighthave** [comment]
I have this mostly implemented, except for streaming.  Using Exoplayer makes sense to me.  I'll make a pull request tomorrow.  You can see my commits here: https://github.com/eighthave/NewPipe/tree/master

**Comment by gjedeer** [comment]
Nice stuff @eighthave, much better than mine

**Comment by moshpirit** [comment]
So can we now watch YouTube videos through Tor? I don't want to give any information to Google (I know we can't trust Tor at 100%, but you know what I mean). 

Now I'm using WebTube.

**Comment by JasperWeiss** [comment]
At the moment your searches and all other data except the video stream can be routed through tor (once you enable it in the settings that is)
Video streams over tor will be available soon

**Comment by moshpirit** [comment]
Cool! thanks! 

In WebTube there's an Orbot library and it's supposed to use Tor as well, should it work stream also on Tor? I know it's a different project, but you probably know about this library and suppose the implementation (I'm just a normal user, sorry)

**Comment by JasperWeiss** [comment]
Yes WebTube uses Tor exlusively (All traffic, including streams). Also note I might have discovered Newpipe isn't properly using Tor so you might want to use WebTube for the time being.

**Comment by moshpirit** [comment]
Cool! then why not just implement the Orbot library that WebTube uses?

**Comment by JasperWeiss** [comment]
(You should tap enable Tor of course, it's under the hovering plus icon at the bottom right)

**Comment by moshpirit** [comment]
>(You should tap enable Tor of course, it's under the hovering plus icon at the bottom right)

Yes, I know, thanks.

**Comment by JasperWeiss** [comment]
I believe @eighthave was working on this. As far as I know the guardian project made a library themselves for Orbot support so I'm guessing that's what WebTube is using. 
I guess NewPipe should two since it doesn't actually use Tor now.

**Comment by eighthave** [comment]
Both NewPipe and WebTube are using our library called NetCipher for Tor
integration.  WebTube is using a WebView, where we had to hack in Tor
support.  That approach has leaks and we have stopped working on that.
NewPipe is using proxied HttpURLConnection, which is officially supported.

**Comment by JasperWeiss** [comment]
@eighthave at the moment it doesn't seem to work properly though. For me NewPipe no longer works when I disable it's internet access meaning it's not using orbots proxy. Can you confirm this?

**Comment by eighthave** [comment]
for me NewPipe searches and plays videos whether I have "Use Tor"
enabled or not.  Can you describe your issue more?

**Comment by JasperWeiss** [comment]
So I have 'Use Tor' enabled. So it should use orbots proxy to access the internet. And I'm using Netguard to allow/deny internet access per app.

When I allow internet access to NewPipe it works fine, obviously.
But when I deny internet access to NewPipe it doesn't load anything even though it should work just fine since it is supposed to use the proxy.

Orfox for example (tor browser for android) works with internet access blocked since it uses orbot.
NewPipe refuses to load anything when Internet access it blocked.

So, this means that somehow, direct internet access (not routed through tor) is used. I know streaming isn't routed through tor but searching should be. Perhaps it first tries to do something over direct internet access and then after that it uses tor. But without direct internet access it won't work.

**Comment by JasperWeiss** [comment]
Aha. So I enabled interner access for NewPipe and looked at orbots network usage.
When you search, it doesn't use Tor. It only uses Tor when loading the video details page.

I can confirm this.
1 allow newpipe internet access
2 search
3 disable newpipe internet access
4 click on a video

Loading de video details page works with internet access dissabled since it uses orbot. But search doesn't.

I can't see orbot network activity while searching videos but I can see orbot being used when clicking on one of the results.

**Comment by eighthave** [comment]
thanks for the testing and reporting.  The Tor support in NewPipe is definitely a work in progress.  We've working on a big new update on the NetCipher library used to make Tor integration easy.  Once that's complete, then we'll come back to the various apps like this and update the NetCipher/Tor support.
